Use new pills UI for `can_manage_group` setting.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First three commits are prep commits with one of them adding a new function to check permission when a setting is set to anonymous group and the other two being code refactors.
- Fourth commit is to add a new module to handle pills for group setting.
- Fifth commit is to not show `role:` for system groups.
- Sixth commit is to add code for typeahead shown for group settings.
- Last commit is to use the pills UI for `can_manage_group` setting.

Part of #28808.
 <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![groups-ui-create](https://github.com/user-attachments/assets/c5f06895-6209-4743-9745-6f74127f97ca)
![groups-ui-update](https://github.com/user-attachments/assets/a32800ab-24a2-4d4d-a0ec-bd76843bbd82)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on web/src/settings_config.ts:
```diff
@@ -919,34 +919,42 @@ export const desktop_icon_count_display_values = {
     {
         name: "role:everyone",
-        display_name: $t({defaultMessage: "Admins, moderators, members and guests"}),
+        dropdown_option_name: $t({defaultMessage: "Admins, moderators, members and guests"}),
```
Maybe we can name this field something like "description".

timabbott replied:
Yeah, we can see; we may find we end up deleting it after migrating all the groups inputs to the modern format, so probably not worth effort on its name at this time.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -672,3 +673,11 @@ export function query_matches_name(
 }
+
+export function query_matches_group_name(query: string, user_group: UserGroupPillData): boolean {
+    return typeahead.query_matches_string_in_order(
```
There are some places where we still use `query_matches_name` for groups, but that would work as those typeahead does not include system groups and the only thing that is different in this is that it considers `display_name` while matching with system groups.

timabbott replied:
Yeah, that seems OK, though it would be better to just have them use this too, to avoid problems with cargo-culting creating broken code.

sahil839 replied:
Added a commit for this.

sahil839 commented on web/src/user_groups.ts:
```diff
@@ -405,14 +405,14 @@ export function get_realm_user_groups_for_dropdown_list_widget(
-// `settings_config` yet.
-export function get_display_group_name(user_group: UserGroup): string {
-    return user_group.name === "role:everyone" ? $t({defaultMessage: "Everyone"}) : user_group.name;
+export function get_display_group_name(group_name: string): string {
```
Accepting `group_name` here instead of `group` because the typeahead item is not a `UserGroup` object.

sahil839 commented on web/src/typeahead_helper.ts:
```diff
@@ -557,6 +558,151 @@ export function sort_recipients<UserType extends UserOrMentionPillData | UserPil
+
+    const no_matches = sort_group_setting_items([
+        ...groups_results.no_matches,
+        ...email_results.no_matches,
```
The typeahead is sorted in following order -
- exact match
- prefix match without any preference given whether the case matches or not.
- word boundary match
- no matches

And the each list above is sorted in order system groups > groups > users who are members of group (for group setting) > other users. We do not have group creator data now (the PR for that is close to being completed), so can add that later.

alya replied:
Playing with the PR, let's drop the exact mach distinction -- I think we should treat matches the same regardless of case.

I'm not sure if we'll end up wanting to prioritize groups more strongly, but it'll be easier to play with that after we make this change.

alya replied:
Showing bots mixed in with users feels off, so let's also put all humans above all bots. We can do:

users who are members of group > other users > bots who are members of the group > other bots

sahil839 replied:
The exact match distinction is not for case, it matches only when entire name of user/group is present in the input like below irrespective whether case matches or not.

![Screenshot from 2024-09-07 12-09-14](https://github.com/user-attachments/assets/9c36c539-912e-4eba-b3f2-ab55defd6cd4)

sahil839 replied:
And we do not show bots at all in the typeahead.

alya replied:
Oh, I got confused somehow. It was seeing the bots mixed up with people in group membership that felt off. We can tackle that separately.

alya replied:
But why do I see this? I would expect Moderators to be ranked higher.

![Screenshot 2024-09-09 at 15 40 45](https://github.com/user-attachments/assets/18f956f4-e8b3-4ee1-8ae9-63ed38e4829e)

sahil839 replied:
Fixed this -
![Screenshot from 2024-09-10 21-39-35](https://github.com/user-attachments/assets/c32bd9ab-4153-4ce5-89df-fe3f25441015)

sahil839 commented on web/e2e-tests/navigation.test.ts:
```diff
@@ -72,7 +72,7 @@ async function test_reload_hash(page: Page): Promise<void> {
     await page.evaluate(() => {
         // We haven't converted reload.js to TypeScript yet.
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+
```
Not sure why linter is changing this everytime I run `./tools/lint`.

sahil839 replied:
The linter on CI failed due to this change, so added it back.

sahil839 commented:
This is how it looks when setting is disabled - with `x` icon removed from the pill and user cannot type anything in the input with typeahead also not being visible.

![Screenshot from 2024-09-04 16-59-01](https://github.com/user-attachments/assets/5fe5bbdd-f194-44ca-8940-6323248bc1da)

sahil839 commented:
A couple of things to note -
- `Nobody` group is still shown when other groups/users are present.
- How to handle the empty input? Currently it just creates a `UserGroup` with no subgroups or members which means no user has permission. But we can disallow setting the input without any pills.

Also, this is just the basic design and the work for guiding users to good structure is left.

alya commented:
This is great! I'll file some ideas from playing with it as follow-up issues.

For "Nobody", I think it might be best to show an empty text field, rather than a pill. We can have some placeholder text:

> "Add roles, groups or users"

alya commented:
Let's rename "Admins" -> "Administrators", since we generally don't abbreviate it.

alya commented:
Let's show the typeahead right away, as soon as you put your cursor into the field, like we do for "Channels they should join" in the invitation modal.

alya commented:
Follow-up; I'll start a #design discussion: #31550

sahil839 commented:
Updated the PR -
- We do not show "Nobody" pill in the container and typehead. And empty container means eventually the setting would be set to "Nobody" group.
- Added placeholder for the input and also the typeahead is shown when input is empty.
- "Full members" group is not shown in typeahead if waiting period is set to 0. But the group can still be added if user types the complete group name, i.e. `role:fullmembers`.
- Renamed "Admins" to "Administrators".

sahil839 commented:
One thing that we do not show certain groups in typeahead like `Nobody` and `Everyone` (which is not allowed for this setting), but those are added to the container if user types the complete name by themselves. But the server will return error for such cases. Do we need to also handle this by not converting those group names to pills?

alya commented:
Yeah, I think the ideal behavior is that anything that returns a server error should not become a pill. It's OK if this is left as a follow-up, as it's unlikely to come up.

alya commented:
Re-testing the PR, when I create a new group (which I didn't test last time):

1. The "who can manager" field looks disabled.
2. The dropdown is slightly shorter than the text fields. It can be a follow-up, but having them line up would look better.

![Screenshot 2024-09-09 at 15 31 15](https://github.com/user-attachments/assets/8c8711ab-9336-41b7-a5b0-5aa3056b580c)

sahil839 commented:
Updated the field width and also its color to not look disabled -
![Screenshot from 2024-09-10 21-40-00](https://github.com/user-attachments/assets/9aeb63b5-4cbe-413e-a087-ec32d0d30917)

sahil839 commented:
Also, added changes to not create pill for a group which is not allowed even if user types the complete name. This includes "Nobody" group and "Full members" group as well if there is no waiting period.

timabbott commented on web/src/user_groups.ts:
```diff
@@ -284,6 +284,28 @@ export function is_user_in_group(user_group_id: number, user_id: number): boolea
+        }
+    }
+    return false;
+}
```
This probably isn't important, but the set-based algorithm using `get_recursive_subgroups` would be more efficient if there were a lot of direct subgroups that shared subgroups.

timabbott commented:
I merged some of the preparatory refactors as #31623 to start shrinking this.

timabbott commented on web/src/user_groups.ts:
```diff
@@ -354,6 +355,14 @@ export function check_system_user_group_allowed_for_setting(
+        for_new_settings_ui &&
+        realm.realm_waiting_period_threshold === 0
+    ) {
+        return false;
```
We probably want a comment explaining that we hide the full members group when it's not defined as separate from members.

sahil839 replied:
Added a comment.

timabbott commented on web/src/user_groups.ts:
```diff
@@ -325,7 +333,7 @@ function get_display_name_for_system_group_option(setting_name: string, name: st
     group_name: string,
     group_setting_config: GroupPermissionSetting,
-    for_new_settings_ui = false,
+    for_new_settings_ui: boolean,
```
`for_new_settings_ui` is probably a name that we should plan to be temporary; "new" loses its meaning over time. But I'm not sure what to replace it with, so it can stay for now..

timabbott commented on web/src/typeahead_helper.ts:
```diff
@@ -555,6 +556,153 @@ export function sort_recipients<UserType extends UserOrMentionPillData | UserPil
+            return -1;
+        }
+
+        return 1;
```
Technically a sorting function should return 0 if the values are the same, right? But I guess that should not happen.

sahil839 replied:
Yes, two groups cannot have same names.

timabbott commented on web/src/settings_org.js:
```diff
@@ -613,8 +613,10 @@ export function discard_group_property_element_changes(elem, group) {
+    if (property_name === "can_mention_group") {
         settings_components.set_dropdown_list_widget_setting_value(property_name, property_value);
+    } else if (property_name === "can_manage_group") {
+        settings_components.set_group_setting_widget_value(property_name, property_value);
```
We'll want to get out of hardcoding lists of elements, and instead doing some item type check. But I guess that'll get easier once we migrate `can_mention_group` too.

sahil839 replied:
There is no other setting in the groups UI, so we can also remove the next else-if block now and will not need any if condition once we migrate `can_mention_group`.

timabbott commented on web/src/settings_components.ts:
```diff
@@ -1348,3 +1391,93 @@ export function initialize_disable_btn_hint_popover(
+        group_setting_widget_map.set("new_group_" + setting_name, pill_widget);
+    } else {
+        group_setting_widget_map.set(setting_name, pill_widget);
+    }
```
I'm not sure this `new_group` pattern is a great one; why does this need to be a different element?

Seems OK to consider cleaning up in a follow-up PR regardless.

sahil839 replied:
We need to store the pill-widget for new group and existing group separately, because the widget for new group is set-up once and is not set-up again while navigating between group edit and group creation UI.

sahil839 replied:
But can see if we can have a better approach as a follow-up.

timabbott replied:
Yeah, what I'm thinking is maybe we wouldn't need this if we just set it up whenever entering the panel...

timabbott commented:
This lgtm but I'm not sure how best to resolve the merge conflict; DM me when you've resolved that so I can merge this to keep ths project moving.

sahil839 commented:
Rebased the PR and posted replies to above comments.

Also, now that we have merged the PR to not allow guest permission when the setting does not allow "Everyone" group, do we want to avoid showing guest users in the typeahead? We do not restrict them in API yet though and they can be other user-defined named groups as well.

timabbott commented:
Yes, we should probably hide guests from the typeahead there. Don't need to do something with groups. I think we don't need to restrict at the API layer; because you can change roles between guest and not, etc., I don't think we can make a hard restriction without undue effort, and we've already done the key work to make it impossible to exercise the permission anyway.
